blockindex/liquidstaking_indexer.go

```diff
@@ -0,0 +1,950 @@
+			"name": "Unpaused",
+			"type": "event"
+		},
+		{
+			"anonymous": false,
+			"inputs": [
+				{
+					"indexed": true,
+					"internalType": "uint256",
+					"name": "tokenId",
+					"type": "uint256"
+				}
+			],
+			"name": "Unstaked",
+			"type": "event"
+		},
+		{
+			"anonymous": false,
+			"inputs": [
+				{
+					"indexed": true,
+					"internalType": "uint256",
+					"name": "tokenId",
+					"type": "uint256"
+				},
+				{
+					"indexed": true,
+					"internalType": "address",
+					"name": "recipient",
+					"type": "address"
+				}
+			],
+			"name": "Withdrawal",
+			"type": "event"
+		}
+	]`
+
+	// bucket related namespace in db
+	_liquidStakingBucketInfoNS = "lsbInfo"
+	_liquidStakingBucketTypeNS = "lsbType"
+	_liquidStakingHeightNS     = "lsHeight"
+)
+
+type (
+	// LiquidStakingIndexer is the interface of liquid staking indexer
+	LiquidStakingIndexer interface {
+		blockdao.BlockIndexer
+
+		CandidateVotes(candidate string) *big.Int
+		Buckets() ([]*Bucket, error)
+		Bucket(id uint64) (*Bucket, error)
+	}
+
+	liquidStakingIndexer struct {
+		dirty      batch.CachedBatch   // batch for dirty data
+		kvstore    db.KVStore          // persistent storage
+		dirtyCache *liquidStakingCache // in-memory index for dirty data
+		cleanCache *liquidStakingCache // in-memory index for clean data
+
+		tokenOwner    map[uint64]string // token id -> owner
+		blockInterval time.Duration
+	}
+)
+
+var (
+	_liquidStakingInterface abi.ABI
+	_liquidStakingHeightKey = []byte("lsHeight")
+
+	errBucketTypeNotExist = errors.New("bucket type does not exist")
+
+	// ErrBucketInfoNotExist is the error when bucket does not exist
+	ErrBucketInfoNotExist = errors.New("bucket info does not exist")
+)
+
+func init() {
+	var err error
+	_liquidStakingInterface, err = abi.JSON(strings.NewReader(LiquidStakingContractABI))
+	if err != nil {
+		panic(err)
+	}
+}
+
+// NewLiquidStakingIndexer creates a new liquid staking indexer
+func NewLiquidStakingIndexer(kvStore db.KVStore, blockInterval time.Duration) LiquidStakingIndexer {
+	return &liquidStakingIndexer{
+		blockInterval: blockInterval,
+		dirty:         batch.NewCachedBatch(),
+		dirtyCache:    newLiquidStakingCache(),
+		kvstore:       kvStore,
+		cleanCache:    newLiquidStakingCache(),
+		tokenOwner:    make(map[uint64]string),
+	}
+}
+
+// Start starts the indexer
+func (s *liquidStakingIndexer) Start(ctx context.Context) error {
+	if err := s.kvstore.Start(ctx); err != nil {
+		return err
+	}
+	return s.loadCache()
+}
+
+// Stop stops the indexer
+func (s *liquidStakingIndexer) Stop(ctx context.Context) error {
+	if err := s.kvstore.Stop(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+// PutBlock puts a block into indexer
+func (s *liquidStakingIndexer) PutBlock(ctx context.Context, blk *block.Block) error {
+	actionMap := make(map[hash.Hash256]*action.SealedEnvelope)
+	for _, act := range blk.Actions {
+		h, err := act.Hash()
+		if err != nil {
+			return err
+		}
+		actionMap[h] = &act
```

Comment:
this won't work. `act` is a local variable, after the `for range` loop, it will be equal to the last item in `blk.Actions[]`
will need to use `actionMap[h] = &blk.Actions[i]`

Reply:
sure, it's one of golang's common mistakes